cpp/tests/utility/vector_equality.hpp

```diff
@@ -69,6 +95,20 @@ MATCHER(float_matcher, std::string(negation ? "are not" : "are") + " approximate
   return false;
 }
 
+MATCHER_P(float_near_matcher,
```

Comment:
The `float_matcher` is also a "near" matcher, so perhaps `float_near_matcher` is not the right name. The only different is that we pass abs_error. GTEST ignores the abs error if it is zero. So why not just augment the existing matchers with default abs_error of zero to avoid all the additional code?

Reply:
Hmm, it's not that simple.. If you default `abs_error` to 0, technically it's going through a different code path as opposed to using `FLOAT_EQ` matcher. `FLOAT_NEAR(abs_error==0) will go to this:

https://github.com/google/googletest/blob/3026483ae575e2de942db5e760cf95e973308dd5/googlemock/include/gmock/gmock-matchers.h#L1687

which will not use the ULP near equal check but completely resort to `abs_err` check.

But `FLOAT_EQ` matcher will use this:

https://github.com/google/googletest/blob/3026483ae575e2de942db5e760cf95e973308dd5/googlemock/include/gmock/gmock-matchers.h#L1706

which will use the ULP near equal check.

While we can default `abs_error` to -1, the default value where gtest use internally to differentiate. I believe that's an implementation detail and it's in nature an invalid input to the matcher.

Reply:
If just for naming, we can call it `float_eq_by_ulp` / `float_eq_by_abs_error`. As for the user facing API `expect_vector_equivalent`, I think we should keep it as is and use overloads for user to differentiate.